Pipeline writing conventions (#99)
sanger-tol/pipelines-website: markdown/usage/gap_pipelines.md

---
title: Genome After-Party pipeline standards
subtitle: This page describes the conventions we follow when developing Genome After-Party pipelines
---

## Input principles

Genome After-Party pipelines define inputs in a consistent manner.
The main principles are that:

- We follow the nf-core standards, of course :wink:
- Pipelines must not rely on Sanger infrastructure or environment to
  run. Pipelines must only use provided data files and metadata values,
  and query public APIs.

> - No _custom_ YAML or JSON file. All parameters can be passed directly

Reviewer: Should we specify a bit more on what `custom` mean?

Author: I've changed the list a bit. My point about YAML/JSON is now at the end of the list. Is it clearer ?

I want to say I don't want custom YAML formats like the ToLA pipelines have

  on the command-line. Parameters may be passed in YAML/JSON via
  Nextflow's native `-params-file` option.

> - Pipelines may accept a "samplesheet" as the `--input` parameter.

Reviewer: ```suggestion
- Pipelines should accept a "samplesheet" as the `--input` parameter.
```

  Samplesheets are used to enumerate input files, input parameters,
  output directories.
- The output directory is controlled via the `--outdir` parameter

>   and all outputs conform to the [output convention](gap_conventions).

Reviewer: Maybe a bit more clear on how to control sub-outputs
```suggestion
  and all outputs conform to the [output convention](gap_conventions). This could be achieved by defining publishDir directives in `conf/modules.config` in the pipeline.
```

- One pipeline run should process a single assembly, passed as the
  `--fasta` parameter. Pipelines may propose a bulk-processing mode
  by allowing multiple input assemblies and output directories to be
  passed through the samplesheet.

> - Parameter names should be consistent across pipelines.

Reviewer: Not related to this documents but should we added a summary doc of pipline parameters (which could actually be the params sheet we are using to track params)

Author: I've added a section with the most common parameters that are shared across pipelines.
Is that useful enough ?

Reviewer: That's much better to help developers in checking the params. I am a bit concerned that when someone develop a pipeline, they would need to go through the exsiting pipelines to check if the param exist.

Author: Fair comment. Ideally this could be a pipelines.tol.sanger.ac.uk page. I even think I saw something like that at some point.

Author: Found it. It's in progress in nf-core: https://github.com/nf-core/website/issues/1251 and https://deploy-preview-4213--nf-core-main-site.netlify.app/pipeline_params
though to be honest, we're still on the old PHP codebase so we couldn't copy the code easily (maybe AI could ... ?)

## Input definition

Historically, samplesheets have had a `sample` column, as in nf-core
pipelines, that contains an identifier for the input file.
`sample` generally doesn't need to be an actual _sample_ identifier.
In particular, `sample` values may contain `/` to structure the outputs in
sub-directories. This is typically used to follow the `${specimen}/${run}`
output directory convention.

When the specimen or run identifiers are explicitly needed, they

> should be requested as such in the samplesheet, rather than being inferred

Reviewer: Maybe just a bit clearer
```suggestion
they should be requested as explicit named columns (e.g. specimen, run) in the samplesheet
```

from the `sample` value, cf the [readmapping](/readmapping/usage)
pipeline.
When such identifiers are optional, or to avoid introducing a breaking
change in a pipeline, the string before the `/` may be used as the
specimen identifier, cf the [variantcalling](/variantcalling/usage)
pipeline.

## Execution control

Software that can choose where temporary files are stored should have
this controlled by a pipeline parameter called `--use_work_dir_as_temp`.
When set to _true_, software should be configured to use the task
directory instead. When set to _false_, software should use the host's
temporary space.
By extension, this `--use_work_dir_as_temp` parameter can be used to
control task's `scratch` directive.

Pipelines should define CPU, memory, and runtime requirements tailored
to each software and inputs. We want to avoid nf-core's generic labels
because they are often a poor fit and lead to excessive resource
wastage.
To signify this, pipelines may remove all `withLabel:` that come with
the nf-core template.

## Outputs

> Most of the conventions are already defined in the [output convention](gap_conventions)

Reviewer: Is it needed to update these link to the right link? Similarly to usage docs of pipelines.

Author: Not sure what you mean. This link works fine (on the pipelines website)

Reviewer: Ah, I got error when I view them on github. Then it should be fine!

as they form a contract with downstream consumers.

However, pipelines don't have to explicitly request all possible identifiers
(`${assembly}`, `${run}`, etc) just for the sake of naming the outputs
correctly. Pipelines should use the "baseName" of the input files
whenever possible and assume that Genome After-Party production
_inputs_ will be named in such a way that the naming convention
is fulfilled all the way through.

For instance, the [variantcalling pipeline](/variantcalling/usage)
takes aligned reads in BAM/CRAM and simply changes the extension to
`.deepvariant.vcf.gz`.
The naming convention applies because in production, it is run on
[readmapping](/readmapping/output) outputs that themselves comply.

Lastly, when a pipeline merges data files, it shall name the
merged dataset `merged_${#}` and list the identifiers of the merged

> data files in a file named `SOURCE.txt`.

Reviewer: ```suggestion
data files in a file named `SOURCE.txt`. This file lists the source run (`run accessions`) one per line.
```

## Testing

All pipelines must define both a `test` and `test_full` profile as per

> the [testing conventions](../contributing/testing).

Reviewer: Maybe move this line up here
```suggestion
the [testing conventions](../contributing/testing). The `test` profile shall be used in the `default.nf.test` test.
```

The `test` profile must work as is, without any additional parameters
or prior environment setup, e.g.

```
nextflow run . -profile singularity,test --outdir results
```

The `test_full` must also work as is, without any additional parameters,
but may require the Sanger farm environment, e.g. for database or
input paths:

```
nextflow run . -profile singularity,test_full --outdir results_full
```

> The `test` profile shall be used in the `default.nf.test` test.

Reviewer: Maybe moved this line up there
```suggestion
```
